src/DataBus/AzureDataBusPersistence.cs

```diff
@@ -14,11 +20,63 @@ protected override void Setup(FeatureConfigurationContext context)
 
-            var dataBus = new BlobStorageDataBus(cloudBlobClient.GetContainerReference(dataBusSettings.Container), dataBusSettings, new AsyncTimer());
+            var container = CreateCloudBlobContainer(dataBusSettings);
+
+            var dataBus = new BlobStorageDataBus(container, dataBusSettings, new AsyncTimer());
 
             context.Container.ConfigureComponent(b => dataBus, DependencyLifecycle.SingleInstance);
         }
+
+        static void ThrowIfConnectionStringAndTokenProviderSpecified(DataBusSettings dataBusSettings)
+        {
+            if (dataBusSettings.RenewalTimeBeforeTokenExpires != null && dataBusSettings.UserProvidedConnectionString)
+            {
+                throw new Exception("More than one authentication method to Azure Service was supplied (using connection string and Managed Identity). Use one method only.");
+            }
+        }
+
+        static CloudBlobContainer CreateCloudBlobContainer(DataBusSettings dataBusSettings)
+        {
+            CloudBlobContainer container;
+
+            // Attempt managed identity identity first
+            if (!string.IsNullOrWhiteSpace(dataBusSettings.StorageAccountName))
+            {
+                var azureServiceTokenProvider = new AzureServiceTokenProvider();
+                var state = (azureServiceTokenProvider, dataBusSettings.RenewalTimeBeforeTokenExpires);
+                var tokenAndFrequency = TokenRenewerAsync(state, CancellationToken.None).GetAwaiter().GetResult();
+                var tokenCredential = new TokenCredential(tokenAndFrequency.Token, TokenRenewerAsync, azureServiceTokenProvider, tokenAndFrequency.Frequency.Value);
+                var storageCredentials = new StorageCredentials(tokenCredential);
+                var containerPath = $"https://{dataBusSettings.StorageAccountName}.blob.core.windows.net/{dataBusSettings.Container}";
+
+                container = new CloudBlobContainer(new StorageUri(new Uri(containerPath)), storageCredentials);
+            }
+            else // fallback to connection string
+            {
+                var cloudBlobClient = CloudStorageAccount.Parse(dataBusSettings.ConnectionString).CreateCloudBlobClient();
+                container = cloudBlobClient.GetContainerReference(dataBusSettings.Container);
+            }
+
+            return container;
+        }
+
+        static async Task<NewTokenAndFrequency> TokenRenewerAsync(object state, CancellationToken token = default)
+        {
+            var (azureServiceTokenProvider, renewal) = (ValueTuple<AzureServiceTokenProvider, TimeSpan>)state;
+
+            // Use the same token provider to request a new token.
+            var result = await azureServiceTokenProvider.GetAuthenticationResultAsync("https://storage.azure.com", cancellationToken: token).ConfigureAwait(false);
+
+            // Renew the token before it expires.
+            var next = (result.ExpiresOn - DateTimeOffset.UtcNow) - renewal;
+            if (next.Ticks < 0)
```

Comment:
How does this work?

Reply:
It's a callback that is invoked by Storage library.